File: http-server-tomcat/src/test/java/io/micronaut/servlet/tomcat/websocket/ChatServerWebSocket.java
package io.micronaut.servlet.tomcat.websocket;

import io.micronaut.context.annotation.Requires;
import io.micronaut.http.context.ServerRequestContext;
import io.micronaut.scheduling.TaskExecutors;
import io.micronaut.scheduling.annotation.ExecuteOn;
import io.micronaut.websocket.WebSocketBroadcaster;
import io.micronaut.websocket.WebSocketSession;
import io.micronaut.websocket.annotation.OnClose;
import io.micronaut.websocket.annotation.OnMessage;
import io.micronaut.websocket.annotation.OnOpen;
import io.micronaut.websocket.annotation.ServerWebSocket;

import java.util.function.Predicate;

@Requires(property = "spec.name", value = "TomcatWebSocketSpec")
@ServerWebSocket("/chat/{topic}/{username}")
@ExecuteOn(TaskExecutors.BLOCKING)
public class ChatServerWebSocket {

    private final WebSocketBroadcaster broadcaster;
    private String openThreadName;
    private String messageThreadName;
    private boolean openThreadVirtual;
    private boolean messageThreadVirtual;

    public ChatServerWebSocket(WebSocketBroadcaster broadcaster) {
        this.broadcaster = broadcaster;
    }

    @OnOpen
    public void onOpen(String topic, String username, WebSocketSession session) {
        this.openThreadName = Thread.currentThread().getName();
        this.openThreadVirtual = Thread.currentThread().isVirtual();
        assert ServerRequestContext.currentRequest().isPresent();
        broadcaster.broadcastSync("[" + username + "] Joined!", isValid(topic, session));
    }

    @OnMessage
    public void onMessage(String topic, String username, String message, WebSocketSession session) {
        this.messageThreadName = Thread.currentThread().getName();
        this.messageThreadVirtual = Thread.currentThread().isVirtual();
        broadcaster.broadcastSync("[" + username + "] " + message, isValid(topic, session));
    }

    @OnClose
    public void onClose(String topic, String username, WebSocketSession session) {
        broadcaster.broadcastSync("[" + username + "] Disconnected!", isValid(topic, session));
    }

    private Predicate<WebSocketSession> isValid(String topic, WebSocketSession session) {
        return s -> s != session && topic.equalsIgnoreCase(s.getUriVariables().get("topic", String.class, null));
    }

    public String getOpenThreadName() {
        return openThreadName;
    }

    public String getMessageThreadName() {
        return messageThreadName;
    }

    public boolean isOpenThreadVirtual() {
        return openThreadVirtual;
    }

    public boolean isMessageThreadVirtual() {
        return messageThreadVirtual;
    }

    /**
     * @return whether the handler ran somewhere it is safe to block, which is what
     * {@code @ExecuteOn(BLOCKING)} guarantees. Micronaut runs the handler inline when the
     * container thread already offers that guarantee, as Tomcat's virtual threads do.
     */
    public static boolean isBlockingSafe(String threadName, boolean virtual) {
        return virtual || threadName.startsWith("virtual-executor") || threadName.contains("io-executor");
    }
}

File: test-suite-python/src/test/python/micronaut/servlet/docs/websocket/ChatServerWebSocket.py
# tag::imports[]
from micronaut.websocket import WebSocketBroadcaster, WebSocketSession
from micronaut.websocket.annotation import OnClose, OnMessage, OnOpen, ServerWebSocket
# end::imports[]

from java.lang import String
from micronaut.context.annotation import Requires


@Requires(property="spec.name", value="ChatWebSocketSpec")
# tag::class[]
@ServerWebSocket("/ws/chat/{topic}/{username}")  # <1>
class ChatServerWebSocket:

    def __init__(self, broadcaster: WebSocketBroadcaster):
        self.broadcaster = broadcaster

    @OnOpen  # <2>
    def on_open(self, topic: str, username: str, session: WebSocketSession) -> None:
        self.broadcaster.broadcastSync("[" + username + "] Joined!", self.is_valid(topic, session))

    @OnMessage  # <3>
    def on_message(self, topic: str, username: str, message: str, session: WebSocketSession) -> None:
        self.broadcaster.broadcastSync("[" + username + "] " + message, self.is_valid(topic, session))  # <4>

    @OnClose  # <5>
    def on_close(self, topic: str, username: str, session: WebSocketSession) -> None:
        self.broadcaster.broadcastSync("[" + username + "] Disconnected!", self.is_valid(topic, session))

    def is_valid(self, topic: str, session: WebSocketSession):
        return lambda s: s != session and topic.lower() == str(s.getUriVariables().get("topic", String, None)).lower()
# end::class[]
